### PR TITLE
senegal

### PR DESCRIPTION
**Pull request by abhidg**
- parsers: add ccp-senegal
- schemas: allow date and start_date to include time
- schemas(observation): add loss of smell or taste

> adtl isaric/parsers/ccp-senegal.toml ../isaric-data/ccp-senegal/CVRZVOV_GLOBAL_Changes.csv

|table       	|valid	|total	|percentage_valid|
|---------------|-------|-------|----------------|
|subject       	|200	|200	|100.000000% |
|visit         	|200	|200	|100.000000% |
|observation   	|25011	|25019	|99.968024% |

## observation

* 6: data must contain ['phase', 'date', 'name'] properties
* 2: data must be valid exactly by one definition (0 matches found)

### CODE REVIEW

**Comment by pipliggins** [comment]
All the drugs administered (antivirals, antibiotics, steroids etc) appear to be stored in tmt_pdhosp_nom{n} variables (of which there are 20ish in the data). Any thoughts about how to include these? We'll have to work backwards from finding names and inferring that antivirals etc. were administered I think.

It's free text data, but appears to have been somewhat cleaned, it's all in caps so as a stop gap we could for example search for 'HYDROCORTISONE' in those fields, and if present then `treatment_corticosteroid` will be True. It's a long way from catching everything but prior to getting NLP in there that might be as good as we can do.

**Comment by abhidg** [comment]
Will have a think, best to programmatically generate these mappings somehow. There are also some names in French and spelling errors.

File: schemas/dev/observation.schema.json [hunk 1]
```diff
@@ -124,6 +138,7 @@
         "joint_pain",
         "loss_of_smell",
         "loss_of_taste",
+        "loss_of_smell_or_taste",
```
Comment (ekamau) [inline schemas/dev/observation.schema.json:141]: nice add! i think one of the recent had taste and smell combined into one field ..

**Review by ekamau — APPROVED** [review]
looks good!

File: isaric/parsers/ccp-senegal.toml [hunk 2]
```diff
@@ -0,0 +1,687 @@
...
+  name = "wheezing"
+  phase = "admission"
+  date = { ref = "admissionDate" }
+  is_present = { field = "wheeze_ceoccur_v2", ref = "Y/N/NK" }
+
+[[observation]]
+  name = "shortness_of_breath"
+  phase = "admission"
+  start_date = { field = "shortbreath_cestdat_v2" }
+  date = { ref = "admissionDate" }
```
[the remark this answers is not in the document]
Comment (abhidg) [reply isaric/parsers/ccp-senegal.toml:627]: Yes - start date is date of symptom onset

File: isaric/parsers/ccp-senegal.toml [hunk 3]
```diff
@@ -0,0 +1,739 @@
...
+    schema = "../../schemas/dev/visit.schema.json"
+
+  [adtl.tables.observation]
+    kind = "oneToMany"
+    schema = "../../schemas/dev/observation.schema.json"
+    common = { visit_id = { field = "record_id", sensitive = true } }
+
+[adtl.defs]
+  "Y/N/NK" = { values = { 1 = true, 2 = false } }
+  "né vivant/mort-né" = { values = { 1 = "né vivant", 2 = "mort-né" } }
```
Comment (pipliggins) [inline isaric/parsers/ccp-senegal.toml:28]: I don't think this is being used - but should make sure text is always returned in a single language (i.e. English for now).

File: isaric/parsers/ccp-senegal.toml [hunk 4]
```diff
@@ -0,0 +1,739 @@
...
+    aggregation = "lastNotNull"
+    schema = "../../schemas/dev/visit.schema.json"
+
+  [adtl.tables.observation]
+    kind = "oneToMany"
+    schema = "../../schemas/dev/observation.schema.json"
+    common = { visit_id = { field = "record_id", sensitive = true } }
+
+[adtl.defs]
+  "Y/N/NK" = { values = { 1 = true, 2 = false } }
```
Comment (pipliggins) [inline isaric/parsers/ccp-senegal.toml:27]: With the pending adtl update, should this be updated to include 3=None ?

File: isaric/parsers/ccp-senegal.toml [hunk 5]
```diff
@@ -0,0 +1,739 @@
...
+  values = { 1 = true, 0 = false }
+
+[subject.icu_admitted]
+  field = "daily_hoterm"
+  ref = "Y/N/NK"
+
+[subject.sex_at_birth]
+  field = "sex"
+  description = "Sex"
+  values = { 1 = "male", 2 = "female", 3 = "non_binary" }
```
Comment (pipliggins) [inline isaric/parsers/ccp-senegal.toml:200]: 3 = Unknown / unspecified rather than non-binary

File: isaric/parsers/ccp-senegal.toml [hunk 6]
```diff
@@ -0,0 +1,739 @@
...
+      3 = "transferred"
+      4 = "death"
+      5 = "palliative discharge"
+
+[visit.date_outcome]
+  field = "dsstdtc"
+  description = "Outcome date"
+
+[visit.treatment_dialysis]
+  field = "daily_rrt_cmtrt"
```
Comment (pipliggins) [inline isaric/parsers/ccp-senegal.toml:242]: add renal_proccur

File: isaric/parsers/ccp-senegal.toml [hunk 7]
```diff
@@ -0,0 +1,739 @@
...
+  name = "glasgow_coma_score"
+  date = { ref = "dailyDate" }
+  phase = "study"
+  value = { field = "daily_gcs_vsorres" }
+
+[[observation]]
+  name = "cough"
+  date = { ref = "dailyDate" }
+  phase = "study"
+  is_present = { field = "daily_cough_ceoccur", ref = "Y/N/NK" }
```
Comment (pipliggins) [inline isaric/parsers/ccp-senegal.toml:358]: If 'cough' is an upper level hierarchy, make combined type with the +sputum/+haemoptysis values that share the same date

File: isaric/parsers/ccp-senegal.toml [hunk 8]
```diff
@@ -0,0 +1,739 @@
...
+      { field = "oxy_vsorresu", values = { 1 = "Room air", 2 = "Oxygen therapy" } },
+    ]
+
+[[observation]]
+  name = "cough"
+  phase = "admission"
+  start_date = { field = "cough_cestdat_v2" }
+  date = { ref = "admissionDate" }
+  duration_type = "event"
+  is_present = { field = "cough_ceoccur_v2", ref = "Y/N/NK" }
```
Comment (pipliggins) [inline isaric/parsers/ccp-senegal.toml:577]: Same here - make combinedType with cough sub-types

File: isaric/parsers/ccp-senegal.toml (hunk 7 above)
Comment (abhidg) [reply isaric/parsers/ccp-senegal.toml:358]: Done!

File: isaric/parsers/ccp-senegal.toml (hunk 8 above)
Comment (abhidg) [reply isaric/parsers/ccp-senegal.toml:577]: Not sure how to combine for multiple dates

File: isaric/parsers/ccp-senegal.toml (hunk 5 above)
Comment (abhidg) [reply isaric/parsers/ccp-senegal.toml:200]: Thanks!

File: isaric/parsers/ccp-senegal.toml (hunk 4 above)
Comment (abhidg) [reply isaric/parsers/ccp-senegal.toml:27]: None is not allowed in TOML, will automatically be None as no mapping will be found for 3

File: isaric/parsers/ccp-senegal.toml (hunk 8 above)
Comment (abhidg) [reply isaric/parsers/ccp-senegal.toml:586]: Will use earliest date once https://github.com/globaldothealth/adtl/issues/62 is added

**Pull request merged**